REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [3]:
import pandas as pd

In [4]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [5]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [6]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [7]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [8]:
bus["business_id"].nunique()

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [9]:
bus[bus['city'] == 'San Francisco']['name'].value_counts().head(5)

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [10]:
type(bus['postal_code'][0])

str

#### Question 3b

What are the unique values of postal_code?

In [11]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [27]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]



In [28]:
bus_valid = bus[bus['postal_code'].isin(validZip)]
bus_valid

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

In [29]:
bus_valid['business_id'].count()

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [30]:
vb[pd.isnull(vb['longitude'])]


business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [31]:
bus_mis['business_id'].count()

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [60]:
new_dataf = vb[['postal_code']]#{'postal_code':validZip, 'null_lon':bus_mis, 'not_null_lon':bus}
new_dataf

postal_code
0          94104
1          94109
2          94104
3          94133
4          94110
...          ...
6310       94103
6311       94127
6312       94116
6313       94112
6314       94110

[5999 rows x 1 columns]

In [71]:
bus_valmis = bus_valid[bus_valid['longitude'].isnull()].groupby('postal_code').agg(null_lon = ('business_id','count'))
bus_valmis

null_lon
postal_code          
94102             196
94103             252
94104              59
94105             100
94107             247
94108              82
94109             146
94110             251
94111             115
94112              59
94114             108
94115              76
94116              35
94117              67
94118             100
94121              55
94122             123
94123              53
94124              88
94127              28
94131              14
94132              65
94133             135
94134              29

In [72]:
bus_valnon = bus_valid.groupby('postal_code').agg(not_null_lon = ('longitude','count'))
bus_valnon

not_null_lon
postal_code              
94102                 262
94103                 306
94104                  74
94105                 126
94107                 204
94108                 142
94109                 237
94110                 334
94111                 167
94112                 131
94114                 122
94115                 162
94116                  58
94117                 130
94118                 143
94121                 106
94122                 147
94123                 116
94124                  76
94127                  43
94131                  34
94132                  69
94133                 282
94134                  45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [73]:
newframe = pd.merge(bus_valmis, bus_valnon, how = 'inner', on = 'postal_code')
newframe

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [94]:
ins

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

In [85]:
len(ins['business_id'].unique())

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [102]:
ins.type.value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [98]:
type(ins['type'][0])

str

In [281]:
ins2 =  ins.groupby('type').filter(lambda x : len(x) > 10)
ins2


business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

In [282]:
ins2.date.value_counts()

20160410    143
20140408    124
20150419    107
20140914    105
20161011     93
           ... 
20150131      1
20151114      1
20150411      1
20161203      1
20161223      1
Name: date, Length: 814, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [283]:
ins2['datetime'] = pd.to_datetime(ins2['date'], format='%Y%m%d')
ins2

business_id  score      date     type   datetime
0               10     82  20160503  routine 2016-05-03
1               10     94  20140729  routine 2014-07-29
2               10     92  20140114  routine 2014-01-14
3               19     94  20160513  routine 2016-05-13
4               19     94  20141110  routine 2014-11-10
...            ...    ...       ...      ...        ...
15425        89515     98  20161103  routine 2016-11-03
15426        89701    100  20161206  routine 2016-12-06
15427        89790     96  20161129  routine 2016-11-29
15428        89803    100  20161215  routine 2016-12-15
15429        89900     93  20161206  routine 2016-12-06

[15429 rows x 5 columns]

In [284]:
ins2['year'] = ins2['datetime'].dt.year
ins2

business_id  score      date     type   datetime  year
0               10     82  20160503  routine 2016-05-03  2016
1               10     94  20140729  routine 2014-07-29  2014
2               10     92  20140114  routine 2014-01-14  2014
3               19     94  20160513  routine 2016-05-13  2016
4               19     94  20141110  routine 2014-11-10  2014
...            ...    ...       ...      ...        ...   ...
15425        89515     98  20161103  routine 2016-11-03  2016
15426        89701    100  20161206  routine 2016-12-06  2016
15427        89790     96  20161129  routine 2016-11-29  2016
15428        89803    100  20161215  routine 2016-12-15  2016
15429        89900     93  20161206  routine 2016-12-06  2016

[15429 rows x 6 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [285]:
print(ins2.year.min())
print(ins2.year.max())
#['Year ' + str(i) for i in range(len(ins2))]

2013
2016


In [286]:
ins2.groupby('year').agg({'business_id': ['count']})

business_id
           count
year            
2013          38
2014        5629
2015        3923
2016        5839

In [287]:
ins3 = ins2.groupby('year').filter(lambda x : len(x) > 50)
ins3

business_id  score      date     type   datetime  year
0               10     82  20160503  routine 2016-05-03  2016
1               10     94  20140729  routine 2014-07-29  2014
2               10     92  20140114  routine 2014-01-14  2014
3               19     94  20160513  routine 2016-05-13  2016
4               19     94  20141110  routine 2014-11-10  2014
...            ...    ...       ...      ...        ...   ...
15425        89515     98  20161103  routine 2016-11-03  2016
15426        89701    100  20161206  routine 2016-12-06  2016
15427        89790     96  20161129  routine 2016-11-29  2016
15428        89803    100  20161215  routine 2016-12-15  2016
15429        89900     93  20161206  routine 2016-12-06  2016

[15391 rows x 6 columns]

In [288]:
ins3.groupby('year').agg({'business_id': ['count']})

business_id
           count
year            
2014        5629
2015        3923
2016        5839

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [289]:
ins_2016 = ins3.loc[ins3['year'] == 2016]
ins_2016

business_id  score      date     type   datetime  year
0               10     82  20160503  routine 2016-05-03  2016
3               19     94  20160513  routine 2016-05-13  2016
6               24     98  20161005  routine 2016-10-05  2016
7               24     96  20160311  routine 2016-03-11  2016
10              31     84  20160816  routine 2016-08-16  2016
...            ...    ...       ...      ...        ...   ...
15425        89515     98  20161103  routine 2016-11-03  2016
15426        89701    100  20161206  routine 2016-12-06  2016
15427        89790     96  20161129  routine 2016-11-29  2016
15428        89803    100  20161215  routine 2016-12-15  2016
15429        89900     93  20161206  routine 2016-12-06  2016

[5839 rows x 6 columns]

In [290]:
ins_2016.groupby('year').agg({'business_id': ['count']})

business_id
           count
year            
2016        5839

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [291]:
bus_2016 = pd.merge(bus, ins_2016, how = 'outer', on = 'business_id')
bus_2016

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7606        90234                ACCESS SFUSD THE ARC   
7607        90236                      PHO DONG HUONG   
7608        90237                    WIN'S RESTAURANT   
7609        90238                    USA POWER MARKET   
7610        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7606                1520 HOWARD ST   San Francisco    CA       94103   
7607             667 MONTEREY BLVD   San Francisco    CA       94127   
7608               3040 TARAVAL ST   San Francisco    CA       94116   
7609                1524 OCEAN AVE   San Francisco    CA       94112   
7610               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  score        date     type  \
0     37.791116 -122.403816  +14154217044   82.0  20160503.0  routine   
1     37.786848 -122.421547  +14157763262   94.0  20160513.0  routine   
2     37.792888 -122.403135  +14156779494   98.0  20161005.0  routine   
3     37.792888 -122.403135  +14156779494   96.0  20160311.0  routine   
4     37.807155 -122.419004           NaN   84.0  20160816.0  routine   
...         ...         ...           ...    ...         ...      ...   
7606        NaN         NaN           NaN    NaN         NaN      NaN   
7607        NaN         NaN  +14154945048    NaN         NaN      NaN   
7608        NaN         NaN  +14158606036    NaN         NaN      NaN   
7609        NaN         NaN  +14153331878    NaN         NaN      NaN   
7610        NaN         NaN  +14158881393    NaN         NaN      NaN   

       datetime    year  
0    2016-05-03  2016.0  
1    2016-05-13  2016.0  
2    2016-10-05  2016.0  
3    2016-03-11  2016.0  
4    2016-08-16  2016.0  
...         ...     ...  
7606        NaT     NaN  
7607        NaT     NaN  
7608        NaT     NaN  
7609        NaT     NaN  
7610        NaT     NaN  

[7611 rows x 14 columns]

In [293]:
bus_2016.head(10)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   
5           45                 CHARLIE'S DELI CAFE   
6           45                 CHARLIE'S DELI CAFE   
7           48                          ART'S CAFE   
8           50                          SUSHI ZONE   
9           54                 RHODA GOLDMAN PLAZA   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
5                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
6                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
7                 747 IRVING ST   San Francisco    CA       94122  37.764013   
8              1815  MARKET ST.   San Francisco    CA       94103  37.771437   
9                  2180 POST ST   San Francisco    CA       94115  37.784626   

    longitude  phone_number  score        date     type   datetime    year  
0 -122.403816  +14154217044   82.0  20160503.0  routine 2016-05-03  2016.0  
1 -122.421547  +14157763262   94.0  20160513.0  routine 2016-05-13  2016.0  
2 -122.403135  +14156779494   98.0  20161005.0  routine 2016-10-05  2016.0  
3 -122.403135  +14156779494   96.0  20160311.0  routine 2016-03-11  2016.0  
4 -122.419004           NaN   84.0  20160816.0  routine 2016-08-16  2016.0  
5 -122.413641  +14156415051   84.0  20160614.0  routine 2016-06-14  2016.0  
6 -122.413641  +14156415051   78.0  20160104.0  routine 2016-01-04  2016.0  
7 -122.465749  +14156657440   94.0  20160630.0  routine 2016-06-30  2016.0  
8 -122.423892  +14156211114   80.0  20160614.0  routine 2016-06-14  2016.0  
9 -122.437734  +14153455060    NaN         NaN      NaN        NaT     NaN

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!